lib/galaxy/datatypes/binary.py

```diff
@@ -291,6 +300,14 @@ def set_meta( self, dataset, overwrite = True, **kwd ):
         dataset.metadata.bam_index = index_file
         # Remove temp file
         os.unlink( stderr_name )
+        # Now use pysam with BAI index to determine additional metadata
+        bam_file = csamtools.Samfile( filename=dataset.file_name, mode='rb', index_filename=index_file.file_name )
+        dataset.metadata.bam_header = bam_file.header
+        dataset.metadata.read_groups = [ read_group['ID'] for read_group in dataset.metadata.bam_header.get( 'RG', [] ) if 'ID' in read_group ]
+        dataset.metadata.reference_names = list( bam_file.references )
+        dataset.metadata.reference_lengths = list( bam_file.lengths )
+        dataset.metadata.sort_order = dataset.metadata.bam_header.get( 'HD', {} ).get( 'SO', None )
+        dataset.metadata.bam_version = dataset.metadata.bam_header.get( 'HD', {} ).get( 'VN', None )
```

Comment:
I'll add a silent try/except around this. It is possible that there could be some pysam issue that might cause an error. Not sure of an exact one, but according to the docs, it is possible that pysam might blowup on non-standard headers -- haven't seen it, but better safe than sorry.